File: build_files/kyth_shared/kyth_shared/system/recovery_status.py
"""Single recovery truth — staged / rollback / quarantine in one view.

Repair page previously asked three places:
  * `bootc.has_staged_update()` / `has_rollback_deployment()` (bootc_status, cached 5s)
  * `update_status.UpdateSnapshot.staged` (watcher JSON, 300s)
  * `boot_health.BootHealthState.quarantined` + `quarantine_reason()` (digest quarantine)

This module is the only place that merges them into a terminal Hub banner.
Presentation stays in `page_repair*`, logic stays here so `bootc rollback`
and `clear-quarantine --digest` retry commands are never out of sync.
"""

from __future__ import annotations
import logging

from dataclasses import dataclass

from kyth_shared.boot_health import read_state as read_boot_health, quarantine_reason
from kyth_shared.system.bootc import has_rollback_deployment, has_staged_update
from kyth_shared.update_status import read_update_snapshot

logger = logging.getLogger(__name__)


@dataclass(frozen=True, slots=True)
class RecoveryStatus:
    has_staged: bool = False
    has_rollback: bool = False
    quarantined_digest: str = ""
    quarantine_detail: str = ""
    watcher_staged: bool = False
    # Ready-to-paste retry for Hub banner
    clear_quarantine_cmd: str = ""


# S2: truth table central — staged × rollback × quarantined → banner/retry
# staged | rollback | quarant. | banner
#   0        0        0        | up-to-date
#   1        *        0        | reboot to apply staged
#   *        1        0        | rollback available
#   *        *        1        | quarantined — clear-quarantine retry
RECOVERY_BANNER: dict[tuple[bool, bool, bool], str] = {
    (False, False, False): "up-to-date",
    (True, False, False): "reboot to apply staged",
    (True, True, False): "reboot to apply staged",
    (False, True, False): "rollback available",
    (False, False, True): "quarantined — clear-quarantine retry",
    (True, False, True): "quarantined — clear-quarantine retry",
    (True, True, True): "quarantined — clear-quarantine retry",
    (False, True, True): "quarantined — clear-quarantine retry",
}


def recovery_banner(status: RecoveryStatus) -> str:
    return RECOVERY_BANNER[(bool(status.has_staged), bool(status.has_rollback), bool(status.quarantined_digest))]


def get_recovery_status() -> RecoveryStatus:
    try:
        has_staged = bool(has_staged_update())
    except Exception:
        has_staged = False
    try:
        has_rollback = bool(has_rollback_deployment())
    except Exception:
        has_rollback = False

    quarantined = ""
    detail = ""
    clear_cmd = ""
    try:
        state = read_boot_health()
        # surface the most recent quarantined digest if any
        if state.quarantined:
            # pick newest by last_failed_at
            newest = max(state.quarantined.values(), key=lambda r: r.last_failed_at)
            quarantined = newest.digest
            detail = quarantine_reason(state, newest.digest) or newest.reason
            if quarantined:
                clear_cmd = f"sudo kyth-boot-health clear-quarantine --digest {quarantined}"
    except Exception:
        logger.debug("handled expected exception", exc_info=True)
        pass

    try:
        snap = read_update_snapshot(max_age=600)
        watcher_staged = bool(snap and snap.staged_digest)
    except Exception:
        watcher_staged = False

    return RecoveryStatus(
        has_staged=has_staged or watcher_staged,
        has_rollback=has_rollback,
        quarantined_digest=quarantined,
        quarantine_detail=detail,
        watcher_staged=watcher_staged,
        clear_quarantine_cmd=clear_cmd,
    )

File: src/kyth-shared-rs/src/system/recovery_status.rs
//! Port of `kyth_shared.system.recovery_status` — staged/rollback/quarantined single view.

use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RecoveryStatus {
    pub has_staged: bool,
    pub has_rollback: bool,
    pub quarantined_digest: String,
    pub quarantine_detail: String,
    pub watcher_staged: bool,
    pub clear_quarantine_cmd: String,
}

pub fn recovery_banner(s: &RecoveryStatus) -> String {
    let key = (s.has_staged, s.has_rollback, !s.quarantined_digest.is_empty());
    match key {
        (false, false, false) => "up-to-date".to_string(),
        (true, false, false) => "reboot to apply staged".to_string(),
        (true, true, false) => "reboot to apply staged".to_string(),
        (false, true, false) => "rollback available".to_string(),
        (_, _, true) => "quarantined — clear-quarantine retry".to_string(),
    }
}

pub fn get_recovery_status() -> RecoveryStatus {
    // Read via probe/cache + boot health — simplified live version:
    // has_staged/has_rollback via deployment_history, quarantined via file existence
    let history = crate::system::deployment_history::deployment_history();
    let has_staged = history.iter().find(|d| d.section=="staged").map(|d| d.available).unwrap_or(false);
    let has_rollback = history.iter().find(|d| d.section=="rollback").map(|d| d.available).unwrap_or(false);
    // quarantine: /var/lib/kyth/boot-health.json contains quarantined digest?
    let mut quarantined = String::new();
    let mut detail = String::new();
    if let Ok(text) = std::fs::read_to_string("/var/lib/kyth/boot-health.json") {
        if let Ok(v) = serde_json::from_str::<serde_json::Value>(&text) {
            if let Some(q) = v.get("quarantined_digest").and_then(|x| x.as_str()) {
                if !q.is_empty() { quarantined = q.to_string(); detail = "boot quarantine active".to_string(); }
            }
        }
    }
    let clear_cmd = if !quarantined.is_empty() { format!("sudo kyth-boot-health clear-quarantine --digest {}", quarantined) } else { String::new() };
    RecoveryStatus { has_staged, has_rollback, quarantined_digest: quarantined, quarantine_detail: detail, watcher_staged: has_staged, clear_quarantine_cmd: clear_cmd }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn banner_staged() {
        let s = RecoveryStatus { has_staged:true, has_rollback:false, quarantined_digest:String::new(), quarantine_detail:String::new(), watcher_staged:true, clear_quarantine_cmd:String::new() };
        assert_eq!(recovery_banner(&s), "reboot to apply staged");
    }
    #[test]
    fn banner_quarantined() {
        let s = RecoveryStatus { has_staged:false, has_rollback:false, quarantined_digest:"abc".to_string(), quarantine_detail:String::new(), watcher_staged:false, clear_quarantine_cmd:String::new() };
        assert_eq!(recovery_banner(&s), "quarantined — clear-quarantine retry");
    }
}
